lib/measurement.js

```diff
@@ -0,0 +1,227 @@
+      }
+
+      return error;
+    });
+}
+
+export function fixMeasurements (stream, source) {
+  return stream
+    .map(({
+      date, parameter, value, unit, averagingPeriod,
+      location, city, country, coordinates,
+      attribution, sourceType, mobile
+    }) => ({
+      date,
+      parameter,
+      value,
+      unit,
+      averagingPeriod,
+
+      location: location || source.location || stream.name,
+      city: city || source.city,
+      country: country || source.country,
+      coordinates,
+
+      attribution,
+      sourceName: source.name,
+      sourceType: sourceType || source.type || 'government',
+      mobile: !!(typeof mobile === 'undefined' ? source.mobile : mobile)
+    }));
+}
+
+/**
+ * Verifies if the measurements stream itself adhere's to requirements.
+ *
+ * @param {DataStream<Measurement>} stream
+ */
+export function verifyMeasurementsStreamInstance (stream) {
+  if (typeof stream !== 'object' && stream instanceof DataStream) {
+    let isValid = false;
+    let failures = { 'no data provided': 1 };
+    return { isValid, failures };
+  }
+  let isValid = !!stream.name;
+  let failures = [];
+
+  return { isValid, failures };
+}
+
+/**
+ * Filter measurements from a measurement stream
+ *
+ * @param {DataStream<Measurement>} stream The measurements stream to prune measurements from
+ * @return {DataStream<Measurement>} A stream pruned of invalid measurement objects, may be empty
+ *                                   and a failures object of aggregated reasons for cause failures
+ */
+export function validateMeasurements (stream, source) {
+  const out = stream.map(
+    async measurement => {
+      let v = validate(measurement, measurementSchema);
+
+      if (v.errors.length === 0) {
+        return measurement;
+      } else {
+        return new MeasurementValidationError(source, {measurement, errors: v.errors});
+      }
+    }
+  );
+
+  return out;
+}
+
+/**
+ * Create a function to ask the adapter for cause, verify the cause and output the ready stream.
+ * @param {Object} source A source object
+ * @param {JetLog} errLog An cause log to stream messages to
+ * @return {MeasurementsObject} Measurements object including the stream
+ */
+export async function getMeasurementsFromSource (source, {dryrun}) {
+  const out = new DataStream();
+
+  try {
+    const adapter = await getAdapterForSource(source);
+    log.debug(`Looking up adapter for source "${source && source.name}"`);
+    if (!adapter) {
+      out.end(new FetchError(ADAPTER_NOT_FOUND, source, null, 0));
+    } else {
+      const rawMeasurements = await getStreamFromAdapter(adapter, source);
+
+      rawMeasurements
+        .use(fixMeasurements, source)
+        .use(validateMeasurements, source)
+        .pipe(out)
+      ;
+    }
+  } catch (e) {
+    out.end(new FetchError(ADAPTER_ERROR, source, e, 0));
+  }
+
+  return createFetchObject(out, source, dryrun);
+}
+
+export function printDataToLog (stream, log) {
```

Comment:
I noticed when running `node index.js -d -s 'Beijing US Embassy' -v` I'm seeing output like 

```
2018-09-10T12:45:18.742Z - verbose:  utc=Mon Sep 10 2018 08:00:00 GMT-0400 (EDT), local=2018-09-10T20:00:00+08:00, parameter=pm25, value=41, unit=µg/m³, value=1, unit=hours, location=Beijing US Embassy, city=Beijing, country=CN, latitude=39.95, longitude=116.47, attribution=[name=StateAir.net, url=http://www.stateair.net/web/post/1/1.html], sourceName=Beijing US Embassy, sourceType=government, mobile=false
```
which is not stringified. For this output, I prefer stringified as it'll convert the UTC time properly which makes it easier to examine (instead of it taking into account your local time zone when it's an object). The output at this code block is being stringified, but this must not be where this output is happening for the verbose logging.

Reply:
Hmm... the "why" is quite simple - the schema says that one should be a Date object, second a String:

```
        "date": {
            "type": "object",
            "additionalProperties": false,
            "required": ["utc", "local"],
            "properties": {
                "utc": {
                    "type": "date"
                },
                "local": {
                    "type": "string",
                    "pattern": "\\d{4}-\\d{2}-\\d{2}T\\d{2}:\\d{2}:\\d{2}(\\+|-)\\d{2}:\\d{2}"
                }
            }
        },
```

So when that's stringified you get different valuies.

I'll check if that doesn't pose any additional issues, but could we simply use two strings here? The database should be ok with that and the strings will be the same - and I'll add mapping for old format.

Reply:
We would, but I simply added a date correction step along the way - so no need to change any adapter code.

In general I'm trying to make no changes at all to __all adapters__. Whatever worked before should work after.